bundles/org.openhab.voice.googlestt/src/main/java/org/openhab/voice/googlestt/internal/GoogleSTTService.java

```diff
@@ -0,0 +1,381 @@
+import org.openhab.core.audio.AudioStream;
+import org.openhab.core.auth.client.oauth2.*;
+import org.openhab.core.common.ThreadPoolManager;
+import org.openhab.core.config.core.ConfigurableService;
+import org.openhab.core.config.core.Configuration;
+import org.openhab.core.voice.*;
+import org.osgi.framework.Constants;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.api.gax.rpc.ClientStream;
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.StreamController;
+import com.google.auth.Credentials;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.OAuth2Credentials;
+import com.google.cloud.speech.v1.*;
+import com.google.protobuf.ByteString;
+
+import io.grpc.LoadBalancerRegistry;
+import io.grpc.internal.PickFirstLoadBalancerProvider;
+
+/**
+ * The {@link GoogleSTTService} class is a service implementation to use Google Cloud Speech-to-Text features.
+ *
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = SERVICE_PID, property = Constants.SERVICE_PID + "=" + SERVICE_PID)
+@ConfigurableService(category = SERVICE_CATEGORY, label = SERVICE_NAME, description_uri = SERVICE_CATEGORY + ":"
+        + SERVICE_ID)
+public class GoogleSTTService implements STTService {
+
+    private static final String GCP_AUTH_URI = "https://accounts.google.com/o/oauth2/auth";
+    private static final String GCP_TOKEN_URI = "https://accounts.google.com/o/oauth2/token";
+    private static final String GCP_REDIRECT_URI = "urn:ietf:wg:oauth:2.0:oob";
+    private static final String GCP_SCOPE = "https://www.googleapis.com/auth/cloud-platform";
+
+    private final Logger logger = LoggerFactory.getLogger(GoogleSTTService.class);
+    private final ScheduledExecutorService executor = ThreadPoolManager.getScheduledPool("OH-voice-googlestt");
+    private final OAuthFactory oAuthFactory;
+    private final ConfigurationAdmin configAdmin;
+
+    private GoogleSTTConfiguration config = new GoogleSTTConfiguration();
+    private @Nullable OAuthClientService oAuthService;
+
+    @Activate
+    public GoogleSTTService(final @Reference OAuthFactory oAuthFactory,
+            final @Reference ConfigurationAdmin configAdmin) {
+        LoadBalancerRegistry.getDefaultRegistry().register(new PickFirstLoadBalancerProvider());
+        this.oAuthFactory = oAuthFactory;
+        this.configAdmin = configAdmin;
+    }
+
+    @Activate
+    protected void activate(Map<String, Object> config) {
+        this.config = new Configuration(config).as(GoogleSTTConfiguration.class);
+        executor.submit(() -> GoogleSTTLocale.loadLocales(this.config.refreshSupportedLocales));
```

Comment:
Just a thought: is it not necessary to be sure that all locales are loaded before going further ?

Reply:
What will happen if recognize is called iwhile they are not yet loaded?

Reply:
I think is that case the framework will tell you your language is not supported, I'm not really using those values, they are just returned in the supported languages call, is done in a similar way by the Google TTS that takes the supported languages from the voices which are also loaded asynchronously.